# Звіт до індивідуального завдання №1 «Обчислення математичних виразів»

**Дисципліна:** «Комп'ютерна техніка, алгоритмізація та програмування»  
**Студент:** ____________________  
**Група:** ____________________  
**Варіант:** 4

## Мета роботи

Відпрацювати застосування умовних конструкцій Python та функцій модуля `math` для обчислення математичних виразів з перевіркою області допустимих значень (ОДЗ), а також виробити навички представлення результатів у формі технічного звіту.

## Постановка задачі

Скласти консольний скрипт, який приймає дійсне значення `x`, контролює виконання умов ОДЗ і обчислює значення функції варіанта 4 додатку А методичних вказівок:

$$
Y = 3^{\cos x} - 7.5x - \sqrt[3]{\sin x - 2^x} + \frac{\operatorname{ctg}^3 x}{3}.
$$

Умови визначеності виразу (допуск $t = 10^{-12}$):

- **Котангенс:** `tan(x) ≠ 0`, адже `ctg x = 1 / tan x`.
- **Кубічний корінь** визначений для будь-якого дійсного підкореневого виразу, тому обмежень на $sin(x) - 2^x$ немає.

## Блок-схема алгоритму

```mermaid
flowchart TD
    A(["Початок"]) --> B["Введення x"]
    B --> C["tan_value = tan(x)"]
    C --> D{"abs(tan_value) > t"}
    D -- Ні --> D1["порушено ОДЗ: ctg(x) не визначений"]
    D1 --> X(["Кінець"])
    D -- Так --> E["radicand = sin(x) - 2^x"]
    E --> F["cube_root = copysign(|radicand|^{1/3})"]
    F --> G["cotangent = 1 / tan_value"]
    G --> H["y = 3^{cos x} - 7.5x - cube_root + cotangent^3 / 3"]
    H --> I["Вивести x та y"]
    I --> X
```

_Примітка:_ кубічний корінь обчислюється через `math.copysign`, щоб зберегти знак підкореневого виразу.

## Лістинг програми

In [8]:
# Імпортуємо модуль math для використання тригонометричних і степеневих функцій
import math

# Визначаємо допуск t для чисельних порівнянь, щоб уникати похибок округлення
t = 1e-12
# Зчитуємо значення аргументу x з клавіатури та одразу перетворюємо його на число з плаваючою комою
x = float(input("Введіть x: "))

# Обчислюємо значення тангенса, яке потрібне для перевірки ОДЗ і знаходження котангенса
tan_value = math.tan(x)

# Перевіряємо, чи не дорівнює tan(x) нулю, щоб уникнути ділення при обчисленні котангенса
if abs(tan_value) <= t:
    # Виводимо попередження про порушення області допустимих значень
    print("порушено ОДЗ: ctg(x) не визначений через tan(x)=0")
else:
    # Обчислюємо підкореневий вираз для кубічного кореня
    radicand = math.sin(x) - (2.0 ** x)
    # Визначаємо кубічний корінь із урахуванням знака підкореневого виразу
    cube_root = math.copysign(abs(radicand) ** (1.0 / 3.0), radicand)
    # Обчислюємо котангенс як обернене значення тангенса
    cotangent = 1.0 / tan_value
    # Знаходимо степеневий доданок 3^(cos x)
    power_term = 3.0 ** (math.cos(x))
    # Складаємо усі компоненти формули та отримуємо значення функції y
    y = power_term - 7.5 * x - cube_root + (cotangent ** 3) / 3.0
    # Виводимо початкове значення аргументу для звітності
    print(f"x = {x}")
    # Виводимо обчислене значення функції y
    print(f"y = {y}")


Введіть x:  4


x = 4.0
y = -26.73860080898475


У цьому скрипті спочатку проводиться перевірка області допустимих значень: тангенс аргументу не повинен дорівнювати нулю, бо інакше котангенс не існує.

Після успішної перевірки обчислюються всі частини цільової формули: степенева частина $3^{cos (x)}$, зміщення $-7.5x$, кубічний корінь від різниці $sin (x)$ та $2^x$ і куб котангенса, поділений на 3.

Підсумковий результат виводиться разом зі значенням $x$.


## Результати роботи програми

Тестові значення підібрані так, щоб уникнути точок, де `tan(x) = 0`. Результати округлено до трьох знаків.

| x | y | Коментар |
|---|---|---|
| 4.0 | -26.739 | ОДЗ дотримано, кубічний корінь обчислюється коректно |
| 0.0 | не визначено | ctg(x) не визначений через tan(x)=0 |

При введенні `x`, що робить `tan(x)=0` (наприклад, `x = 0` або `x = π`), програма припиняє обчислення та виводить повідомлення про порушення ОДЗ.

## Висновки

Розроблений скрипт демонструє використання умовних операторів та функцій бібліотеки `math` для обчислення складних виразів із контролем ОДЗ. Підхід забезпечує коректне поводження з кубічними коренями та котангенсом, дозволяючи уникнути арифметичних виключень і тиражувати рішення для різних значень аргументу `x`.
